subsys/usb/class/netusb/function_rndis.c

```diff
@@ -0,0 +1,1000 @@
+NET_BUF_POOL_DEFINE(rndis_cmd_pool, CONFIG_RNDIS_CMD_BUF_COUNT,
+		    CONFIG_RNDIS_CMD_BUF_SIZE, 0, NULL);
+static struct k_fifo rndis_cmd_queue;
+
+struct k_delayed_work notify_work;
+
+/*
+ * Stack for cmd thread
+ */
+static K_THREAD_STACK_DEFINE(cmd_stack, 1024);
+static struct k_thread cmd_thread_data;
+
+static struct __rndis {
+	u8_t mac[6];
+	u32_t net_filter;
+
+	struct net_if *iface;
+	u8_t ep_in;
+	u8_t ep_notify;
+
+	enum {
+		UNINITIALIZED,
+		INITIALIZED,
+	} state;
+
+	u8_t media_status;
+
+	struct net_pkt *in_pkt;	/* Pointer to pkt assembling at the moment */
+	int in_pkt_len;		/* Packet length to be assembled */
+	int skip_bytes;		/* In case of low memory, skip bytes */
+
+	u16_t mtu;
+	u16_t speed; /* TODO: Calculate right speed */
+
+	/* Stat */
+	u32_t rx_err;
+	u32_t tx_err;
+	u32_t rx_no_buf;
+
+	atomic_t notify_count;
+} rndis = {
+	.mac =  { 0x00, 0x11, 0x22, 0x33, 0x44, 0x66 },
+	.mtu = 1500, /* FIXME ? */
+	.media_status = RNDIS_OBJECT_ID_MEDIA_DISCONNECTED,
+	.state = UNINITIALIZED,
+	.skip_bytes = 0,
+};
+
+static u8_t manufacturer[] = { 't', 'e', 's', 't' };
+static u32_t drv_version = 1;
+
+static u32_t object_id_supported[] = {
+	RNDIS_OBJECT_ID_GEN_SUPP_LIST,
+	RNDIS_OBJECT_ID_GEN_HW_STATUS,
+	RNDIS_OBJECT_ID_GEN_SUPP_MEDIA,
+	RNDIS_OBJECT_ID_GEN_IN_USE_MEDIA,
+
+	RNDIS_OBJECT_ID_GEN_MAX_FRAME_SIZE,
+	RNDIS_OBJECT_ID_GEN_LINK_SPEED,
+	RNDIS_OBJECT_ID_GEN_BLOCK_TX_SIZE,
+	RNDIS_OBJECT_ID_GEN_BLOCK_RX_SIZE,
+
+	RNDIS_OBJECT_ID_GEN_VENDOR_ID,
+	RNDIS_OBJECT_ID_GEN_VENDOR_DESC,
+	RNDIS_OBJECT_ID_GEN_VENDOR_DRV_VER,
+
+	RNDIS_OBJECT_ID_GEN_PKT_FILTER,
+	RNDIS_OBJECT_ID_GEN_MAX_TOTAL_SIZE,
+	RNDIS_OBJECT_ID_GEN_CONN_MEDIA_STATUS,
+	RNDIS_OBJECT_ID_GEN_PHYSICAL_MEDIUM,
+#if defined(USE_RNDIS_STATISTICS)
+	/* Using RNDIS statistics puts heavy load on
+	 * USB bus, disable it for now
+	 */
+	RNDIS_OBJECT_ID_GEN_TRANSMIT_OK,
+	RNDIS_OBJECT_ID_GEN_RECEIVE_OK,
+	RNDIS_OBJECT_ID_GEN_TRANSMIT_ERROR,
+	RNDIS_OBJECT_ID_GEN_RECEIVE_ERROR,
+	RNDIS_OBJECT_ID_GEN_RECEIVE_NO_BUF,
+#endif /* USE_RNDIS_STATISTICS */
+	RNDIS_OBJECT_ID_802_3_PERMANENT_ADDRESS,
+	RNDIS_OBJECT_ID_802_3_CURR_ADDRESS,
+	RNDIS_OBJECT_ID_802_3_MCAST_LIST,
+	RNDIS_OBJECT_ID_802_3_MAX_LIST_SIZE,
+	RNDIS_OBJECT_ID_802_3_MAC_OPTIONS,
+};
+
+static int parse_rndis_header(const u8_t *buffer, u32_t buf_len)
+{
+	struct rndis_payload_packet *hdr = (void *)buffer;
+	u32_t len;
+
+	if (buf_len < sizeof(*hdr)) {
+		SYS_LOG_ERR("Too small packet len %u", buf_len);
+		return -EINVAL;
+	}
+
+	if (hdr->type != sys_cpu_to_le32(RNDIS_DATA_PACKET)) {
+		SYS_LOG_ERR("Wrong data packet type 0x%x",
+			    sys_le32_to_cpu(hdr->type));
+		return -EINVAL;
+	}
+
+	len = sys_le32_to_cpu(hdr->len);
+
+	if (len < sys_le32_to_cpu(hdr->payload_offset) +
+	    sys_le32_to_cpu(hdr->payload_len) +
+	    /*
+	     * Calculate additional offset since payload_offset is calculated
+	     * from the start of itself ;)
+	     */
+	    (u8_t *)&hdr->payload_offset - (u8_t *)hdr) {
```

Comment:
If I'm understanding correctly, `payload_offset` came directly from the wire and it has not been properly validated.  This could lead to problems if the header is malformed, maliciously or not.

Also, the type of a pointer difference is `ptrdiff_t`, a signed integral type; len is an unsigned integral type; will this work in all occasions?

Reply:
NO! This is offset of the payload_offset field from the start of the structure hdr.